card/lib/card/stage_director.rb

```diff
@@ -0,0 +1,235 @@
+  # (for example as left_id or as type_id) and the subcard doesn't
+  # have an id yes (because it gets created in the same act)
+  # then the subcard's store stage is executed before the supercard's store
+  # stage
+  class StageDirector
+    include Stage
+
+    attr_accessor :prior_store, :act, :card, :stage, :parent, :main,
+                  :subdirectors
+
+    def initialize card, opts={}, main=true
+      @card = card
+      @card.director = self
+      @card.prepare_for_phases
+      @stage = nil
+      @running = false
+      @parent = opts[:parent]
+      # has card to be stored before the supercard?
+      @prior_store = opts[:priority]
+      @main = main
+      @subdirectors = SubdirectorArray.initialize_with_subcards(self)
+      register
+    end
+
+    def register
+      Card::DirectorRegister.add self
+    end
+
+    def unregister
+      @card.director = nil
+      @subdirectors = nil
+      @stage = nil
+      @action = nil
+      Card::DirectorRegister.delete self
+    end
+
+    def validation_phase
+      @running = true
+      run_single_stage :initialize
+      run_single_stage :prepare_to_validate
+      run_single_stage :validate
+      @card.expire_pieces if @card.errors.any?
+      @card.errors.empty?
+    end
+
+    # everything in here can use dirty marks
+    def storage_phase &block
+      # for a subcard :prepare_to_store was already executed
+      # don't execute it twice
+      catch_up_to_stage :prepare_to_store
+      run_single_stage :store, &block
+      run_single_stage :finalize
+    ensure
+      @from_trash = nil
+    end
+
+    # dirty marks are gone in this phase
+    def integration_phase
+      run_single_stage :integrate
+      run_single_stage :integrate_with_delay
+    ensure
+      unregister
+    end
+
+    def catch_up_to_stage next_stage
+      @stage ||= -1
+      (@stage + 1).upto(stage_index(next_stage)) do |i|
+        run_single_stage stage_symbol(i)
+      end
+    end
+
+    def call_after_store &block
+      @call_after_store ||= []
+      @call_after_store << block
+    end
+
+    def need_act
+      act_director = main_director
+      if !act_director
+        fail Card::Error, 'act requested without a main stage director'
+      end
+      act_director.act ||=  Card::Act.create(ip_address: Env.ip)
+      @card.current_act = @act = act_director.act
+    end
+
+    def main_director
+      if main?
+        self
+      else
+        DirectorRegister.act_director || (@parent && @parent.main_director)
+      end
+    end
+
+    def main?
+      @main
+    end
+
+    def running?
+      @running
+    end
+
+    private
+
+    def run_single_stage stage, &block
+      # puts "#{@card.name}: #{stage} stage".red
+      return if @card.errors.any?
+      @stage = stage_index stage
+
+      # in the store stage it can be necessary that
+      # other subcards must be saved before we save this card
+      if stage == :store
+        store &block
```

Comment:
Ambiguous block operator. Parenthesize the method arguments if it's surely a block operator, or add a whitespace to the right of the `&` if it should be a binary AND.